src/hw/raspi4/yield.S

```diff
@@ -0,0 +1,267 @@
+  CALL_KERNEL do_scheduler   // This will set next_thread
+  ldr x6, [x21]              // Get new next_thread
+
+actually_load_thread:
+  /* Either we ran the code above, or kernel jumped here
+     to run setup_thread. Either way we want to set EL0_SP */
+  msr SPSel, #0            // Back to user stack
+  str x6, [x20]            // current_thread = next_thread
+  mov x12, #0              // Set next to null for next switch to call scheduler
+  str x12, [x21]           // next_thread = NULL
+  ldr x20, [x20]           // x20 = current_thread
+  ldr x3, [x20], #8        // x3 = current_thread->stack_ptr
+  mov sp, x3
+
+  /* Restore all registers of the new thread
+     (even init state threads have a blank context
+      to restore)
+  */
+  ldp x29, x30, [sp], #16
+  ldp x27, x28, [sp], #16
+  ldp x25, x26, [sp], #16
+  ldp x23, x24, [sp], #16
+  ldp x21, x22, [sp], #16
+  ldp x19, x20, [sp], #16
+  ldp x17, x18, [sp], #16
+  ldp x15, x16, [sp], #16
+  ldp x13, x14, [sp], #16
+  ldp x11, x12, [sp], #16
+  ldp x9,  x10, [sp], #16
+  ldp x7,  x8,  [sp], #16
+  ldp x5,  x6,  [sp], #16
+  ldp x3,  x4,  [sp], #16
+  ldp x1,  x2,  [sp], #16
+
+  /* x1 = restore PC */
+  msr ELR_EL1, x1
+
+  /* This is FPSR/SPSR */
+  ldp x0, x1, [sp], #16
+  msr FPSR, x0
+  msr SPSR_EL1, x1
+
+  /* Actual x0 and x1 */
+  ldp x0, x1, [sp], #16
+
+  eret
+
+.global signal_handler_wrapper
+signal_handler_wrapper:
+  // x0 = signal, x1 = handler
+  blr x1
+  svc svc_thread_switch
+.global signal_handler_wrapper_end
+signal_handler_wrapper_end:
+  nop
+
+.global get_context
+get_context:
+  // x0 = destination pointer
+  stp x19, x20, [x0], #16
+  stp x21, x22, [x0], #16
+  stp x23, x24, [x0], #16
+  stp x25, x26, [x0], #16
+  stp x27, x28, [x0], #16
+  stp x29, x30, [x0], #16
+  mov x1, sp
+  // Default pc is current lr
+  stp x1, x30, [x0], #16
+  ret
+
+.global set_context
+set_context:
+  // x0 = source pointer
+  ldp x19, x20, [x0], #16
+  ldp x21, x22, [x0], #16
+  ldp x23, x24, [x0], #16
+  ldp x25, x26, [x0], #16
+  ldp x27, x28, [x0], #16
+  ldp x29, x30, [x0], #16
+  // pc and sp
+  ldp x1,  x2,  [x0], #16
+  mov sp, x1
+  br x2
+
+.global set_context_from_stack_address
+set_context_from_stack_address:
+  ldp x0, x1, [sp], #16
+  b set_context
+
+.macro SWAP reg1, reg2
+  ldp x1, x2, [x0]
+  stp \reg1, \reg2, [x0], #16
+  mov \reg1, x1
+  mov \reg2, x2
+.endm
+
+.global swap_context
+swap_context:
+  // x0 = location of ctx to swap to
+  SWAP x19, x20
+  SWAP x21, x22
+  SWAP x23, x24
+  SWAP x25, x26
+  SWAP x27, x28
+
+  // Load x29, x30/lr
+  ldp x3, x4, [x0]
+  // Store current
+  stp x29, x30, [x0], #16
+  mov x29, x3
+  // Don't move new lr just yet
+
+  // Load new sp/pc
+  ldp x1, x2, [x0]
+  mov x3, sp
+  // Store current sp/lr (used as the new pc)
+  stp x3, x30, [x0], #16
+  // Now we can set the new lr
+  mov x30, x4
+  // Set new sp
+  mov sp, x1
+
+  // Reset x0 to be arg1 of new function
+  sub x0, x0, #14*8
+  br x2
```

Comment:
I'm assuming this is an exact copy of the aarch64_virt file (which is fine, code sharing is awkward with the current structure).

Please make sure that it is identical, if not note any divergences.